File: internal/simple/src/sha1.cc
#include "sha1.h"
#include <cassert>

static std::array<uint8_t, 64> PADDING = {
    0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

static uint32_t __attribute__((noinline)) rol32(uint32_t x, uint32_t n) {
  return (x << n) ^ (x << (32 - n));
}

Sha1::Sha1() {}

Sha1::Sha1(std::span<uint8_t> data) { update(data); }

void Sha1::update(std::span<uint8_t> data) {
  std::array<uint32_t, 16> values;
  unsigned int offset = size % 64;
  size += data.size();

  for (auto c : data) {
    buf[offset++] = c;

    if (offset % 64 == 0) {
      for (size_t j = 0; j < 16; ++j) {
        values[j] = (uint32_t)(buf[(j * 4) + 0]) << 24 |
                    (uint32_t)(buf[(j * 4) + 1]) << 16 |
                    (uint32_t)(buf[(j * 4) + 2]) << 8 |
                    (uint32_t)(buf[(j * 4 + 3)]);
      }
      update_values(values);
      offset = 0;
    }
  }
}

void Sha1::update_values(const std::array<uint32_t, 16> &values) {
  std::array<uint32_t, 80> w;
  for (size_t i = 0; i < 16; i++) {
    w[i] = values[i];
  }
  for (size_t i = 16; i < 80; i++) {
    w[i] = rol32(w[i - 3] ^ w[i - 8] ^ w[i - 14] ^ w[i - 16], 1);
  }

  uint32_t a = state[0];
  uint32_t b = state[1];
  uint32_t c = state[2];
  uint32_t d = state[3];
  uint32_t e = state[4];

  for (size_t i = 0; i < 80; i++) {
    uint32_t f;
    uint32_t k;
    if (i < 20) {
      f = (b & c) | ((~b) & d);
      k = 0x5A827999;
    } else if (i < 40) {
      f = b ^ c ^ d;
      k = 0x6ED9EBA1;
    } else if (i < 60) {
      f = (b & c) | (b & d) | (c & d);
      k = 0x8F1BBCDC;
    } else {
      f = b ^ c ^ d;
      k = 0xCA62C1D6;
    }

    uint32_t temp = rol32(a, 5) + f + e + k + w[i];
    e = d;
    d = c;
    c = rol32(b, 30);
    b = a;
    a = temp;
  }
  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;
  state[4] += e;
}

std::array<uint32_t, 5> Sha1::digest_values() && {
  std::array<uint32_t, 16> values;
  size_t message_size = size;
  size_t offset = size % 64;
  size_t padding_length = offset < 56 ? 56 - offset : (56 + 64) - offset;

  update(std::span<uint8_t>(PADDING).subspan(0, padding_length));
  ;

  assert(size % 64 == 56);
  for (unsigned int j = 0; j < 14; ++j) {
    values[j] = (uint32_t)(buf[(j * 4) + 0]) << 24 |
                (uint32_t)(buf[(j * 4) + 1]) << 16 |
                (uint32_t)(buf[(j * 4) + 2]) << 8 |
                (uint32_t)(buf[(j * 4) + 3]);
  }
  values[14] = (uint32_t)((message_size * 8) >> 32);
  values[15] = (uint32_t)(message_size * 8);

  update_values(values);
  return state;
}

std::array<uint8_t, 20> Sha1::digest() && {
  auto state = std::move(*this).digest_values();

  std::array<uint8_t, 20> res;

  for (size_t i = 0; i < 5; i++) {
    res[i * 4 + 0] = ((uint8_t)(state[i] >> 24)) & 0xff;
    res[i * 4 + 1] = ((uint8_t)(state[i] >> 16)) & 0xff;
    res[i * 4 + 2] = ((uint8_t)(state[i] >> 8)) & 0xff;
    res[i * 4 + 3] = ((uint8_t)(state[i] >> 0)) & 0xff;
  }

  return res;
}

File: sploits/simple/sha1.py
from typing import List, Self, Optional

def rol32(x: int, n: int) -> int:
  x %= 2 ** 32
  n %= 32
  return (x << n) ^ (x << (32 - n))

class Sha1:
  PADDING: List[int] = [
    0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
  ]
  size: int
  buf: bytearray


  def __init__(self, data: Optional[bytes]=None):
    self.size = 0
    self.buf = bytearray(64)
    self.state = [
      0x67452301,
      0xEFCDAB89,
      0x98BADCFE,
      0x10325476,
      0xC3D2E1F0,
    ]

    if data is not None:
      self.update(data)

  def update(self, data: bytes):
    values = [None for _ in range(16)]
    offset = self.size % 64
    self.size += len(data)

    for c in data:
      self.buf[offset] = c
      offset += 1
      if offset % 64 == 0:
        for i in range(16):
          values[i] = (self.buf[(i * 4) + 0]) << 24 |\
            (self.buf[(i * 4) + 1]) << 16 |\
            (self.buf[(i * 4) + 2]) << 8 |\
            (self.buf[(i * 4) + 3])
        self.update_values(values)
        offset = 0

  def update_values(self, values: List[int]):
    w = [None for _ in range(80)]
    for i in range(16):
      w[i] = values[i]
    for i in range(16, 80):
      w[i] = rol32(w[i - 3] ^ w[i - 8] ^ w[i - 14] ^ w[i - 16], 1)

    a = self.state[0]
    b = self.state[1]
    c = self.state[2]
    d = self.state[3]
    e = self.state[4]
    for i in range(80):
      if i < 20:
        f = (b & c) | ((~b) & d)
        k = 0x5A827999
      elif i < 40:
        f = b ^ c ^ d
        k = 0x6ED9EBA1
      elif i < 60:
        f = (b & c) | (b & d) | (c & d)
        k = 0x8F1BBCDC
      else:
        f = b ^ c ^ d
        k = 0xCA62C1D6

      temp = rol32(a, 5) + f + e + k + w[i]
      e = d
      d = c
      c = rol32(b, 30)
      b = a
      a = temp

      a = a % 2 ** 32
      b = b % 2 ** 32
      c = c % 2 ** 32
      d = d % 2 ** 32
      e = e % 2 ** 32
    self.state[0] += a
    self.state[1] += b
    self.state[2] += c
    self.state[3] += d
    self.state[4] += e
    self.state = [i % 2 ** 32 for i in self.state]

  def digest_values(self) -> List[int]:
    values = [None for _ in range(16)]
    message_size = self.size
    offset = self.size % 64
    padding_length = 56 - offset if offset < 56 else (56 + 64) - offset
    self.update(self.PADDING[0 : padding_length])

    assert self.size % 64 == 56
    for i in range(14):
      values[i] = (self.buf[(i * 4) + 0]) << 24 |\
        (self.buf[(i * 4) + 1]) << 16 |\
        (self.buf[(i * 4) + 2]) << 8 |\
        (self.buf[(i * 4) + 3])
    values[14] = ((message_size * 8) >> 32) %  2 ** 32
    values[15] = (message_size * 8) % 2 ** 32
    self.update_values(values)
    return self.state

  def digest(self) -> bytes:
    state = self.digest_values()
    res = bytearray(20)

    for i in range(5):
      res[i * 4 + 0] = ((state[i] >> 24)) & 0xff
      res[i * 4 + 1] = ((state[i] >> 16)) & 0xff
      res[i * 4 + 2] = ((state[i] >> 8)) & 0xff
      res[i * 4 + 3] = ((state[i] >> 0)) & 0xff

    return bytes(res)
